File: catalyst_v1/catalyst-api/catalyst_api/schemas/simulation.py
from pydantic import BaseModel, Field
from typing import Dict, Any, Optional


# ============================================================
# Request
# ============================================================

class SimulationRequest(BaseModel):
    risk_reduction_pct: float = Field(
        ...,
        ge=0,
        le=100,
        description="Expected percentage reduction in attrition risk",
        example=25,
    )

    intervention_cost: float = Field(
        ...,
        ge=0,
        description="Estimated annual cost of the intervention",
        example=120000,
    )


# ============================================================
# CFO Impact
# ============================================================

class CFOImpact(BaseModel):
    intervention_cost: float
    cost_avoided: float
    net_roi: float
    roi_multiple: Optional[float]


# ============================================================
# Confidence Band
# ============================================================

class ConfidenceBand(BaseModel):
    low: float
    high: float
    confidence_level: float = Field(
        ...,
        description="Confidence level for the estimated range",
        example=0.8,
    )


# ============================================================
# Response
# ============================================================

class SimulationResponse(BaseModel):
    baseline_cost: float
    simulated_cost: float
    avoided_cost: float
    risk_reduction_pct: float

    cfo_impact: CFOImpact
    confidence: ConfidenceBand

    simulated_kpis: Dict[str, Any]
    diagnostics: Dict[str, Any]

File: catalyst_v1/catalyst-api/catalyst-ui/src/api/simulation.ts
export type CFOImpact = {
  intervention_cost: number;
  cost_avoided: number;
  net_roi: number;
  roi_multiple: number | null;
};

export type ConfidenceBand = {
  low: number;
  high: number;
  confidence_level: number;
};

export type SimulationResponse = {
  baseline_cost: number;
  simulated_cost: number;
  avoided_cost: number;
  risk_reduction_pct: number;

  cfo_impact: CFOImpact;
  confidence: ConfidenceBand;

  simulated_kpis: Record<string, any>;
  diagnostics: Record<string, any>;
};

export async function runSimulation(payload: {
  risk_reduction_pct: number;
  intervention_cost: number;
}): Promise<SimulationResponse> {
  const res = await fetch("http://127.0.0.1:8000/intelligence/simulate", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    throw new Error("Simulation request failed");
  }

  return res.json();
  // -----------------------------
  // Catalyst Simulation Types
  // -----------------------------

}
export type FinancialBand = {
  mean: number;
  p10: number;
  p90: number;
  p2_5: number;
  p97_5: number;
};

export type CatalystSimulationResponse = {
  anchors: {
    peak_psi_quarter: number;
    capital_trough_quarter: number;
    cost_stabilization_quarter: number;
  };

  margin_band: FinancialBand;
  ebitda_band: FinancialBand;
  peak_psi_band: FinancialBand;
  capital_trough_band: FinancialBand;

  steady_state_workforce: number;
};
// -----------------------------
// Catalyst Simulation Call
// -----------------------------

export async function runCatalystSimulation(payload: {
  workforce_size: number;
  avg_cost_per_employee: number;
  revenue_base: number;
  margin_base: number;
  ai_exposure: number;
  leadership_readiness: number;
  margin_buffer: number;
  capital_buffer: number;
  governance_mode: "DEFENSIVE" | "BALANCED" | "AGGRESSIVE";
}): Promise<CatalystSimulationResponse> {
  const res = await fetch("http://127.0.0.1:8000/catalyst/simulate", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      ...payload,
      simulation_context: "DEMO",
    }),
  });

  if (!res.ok) {
    throw new Error("Catalyst simulation request failed");
  }

  return res.json();
}
